File: day3/day3.js
const fs = require('fs');
const input = fs.readFileSync("day3/day3.txt", 'utf-8');

function part1() {
    var x = 0, y = 0;
    var res = new Set();
    res.add("0, 0");
    for (var i = 0; i < input.length; i++) {
        if (input[i] == '^') {
            y++;
        } else if (input[i] == 'v') {
            y--;
        } else if (input[i] == '>') {
            x++;
        } else if (input[i] == '<') {
            x--;
        }
        res.add(x + ", " + y);
    }
    return res.size;
}

function part2() {
    var x = [0, 0], y = [0, 0];
    var res = new Set();
    res.add("0, 0");
    for (var i = 0; i < input.length; i++) {
        if (input[i] == '^') {
            y[i % 2]++;
        } else if (input[i] == 'v') {
            y[i % 2]--;
        } else if (input[i] == '>') {
            x[i % 2]++;
        } else if (input[i] == '<') {
            x[i % 2]--;
        }
        res.add(x[i % 2] + ", " + y[i % 2]);
    }
    return res.size;
}

console.log("Part 1: " + part1());
console.log("Part 2: " + part2());

File: day3/day3.py
directions = open('day3/day3.txt').read()

def part1():
    homes = {(0, 0)}
    x, y = 0, 0
    for d in directions:
        if d == '^':
            y += 1
        elif d == 'v':
            y -= 1
        elif d == '>':
            x += 1
        elif d == '<':
            x -= 1
        homes.add((x, y))
    return len(homes)

def part2():
    homes = {(0, 0)}
    x = [0, 0]
    y = [0, 0]
    for i, d in enumerate(directions):
        if d == '^':
            y[i % 2] += 1
        elif d == 'v':
            y[i % 2] -= 1
        elif d == '>':
            x[i % 2] += 1
        elif d == '<':
            x[i % 2] -= 1
        homes.add((x[i % 2], y[i % 2]))
    return len(homes)

print(f"Part 1: {part1()}")
print(f"Part 2: {part2()}")

# other interesting solution seen on Reddit from u/ilmale:

# def update_pos(c, p):
#     return {'>': (p[0] + 1, p[1]), '<': (p[0] - 1, p[1]), '^': (p[0], p[1] + 1), 'v': (p[0], p[1] - 1)}[c]

# content = directions
# pos = [(0, 0), (0,0)]
# visited_houses = { (0,0): True }
# for i in range( len(content)):
#     t = pos[i%2] = update_pos(content[i], pos[i%2])
#     visited_houses[t] = True
# print(len(visited_houses))